### PR TITLE
Shell: Add support for ARGV (and $*, $#)

### LINKED ISSUE #2948: Shell: `echo $*` is surprising
So I was playing around with the Shell in the Terminal, and:

```
courage ~ $ echo $*
Browser.ini: Command not found

courage ~ $
```

This also happens if the command is executed from a file:

```
courage ~ $ ./foo.sh
Browser.ini: Command not found

courage ~ $
```

I'm not sure what *precisely* `$*` is supposed to do, but ignoring echo and trying to run a file is probably a bug :)

CC @alimpfard

### PR DESCRIPTION
**Pull request by alimpfard**
This patchset also adds the 'shift' builtin, as well as the usual tests.
closes #2948.

It should be noted that this does not adopt the '`$*` is a string, and `$@` is a list` idea, `$*` is a list and `$@` is currently not defined.

more gritty details:
- `$*` is an alias for `$ARGV`, which can be redefined by the user.
- `$#` is an alias for the size of `$ARGV` (we do not yet support any way of checking for list length inside the shell)
- the `shift` builtin also works on `$ARGV`.
- a non-list `$ARGV` is treated as a list containing that only that element.

cc @bugaevc (if you have any input on this)

### CODE REVIEW

File: Shell/main.cpp [hunk 1]
```diff
@@ -181,6 +183,13 @@ int main(int argc, char** argv)
         run_rc_file(Shell::local_init_file_path);
     }
 
+    {
+        Vector<String> args;
+        for (auto* arg : script_args)
+            args.empend(arg);
+        shell->set_local_variable("ARGV", *new AST::ListValue(move(args)));
```
Comment (awesomekling) [inline Shell/main.cpp:190]: This looks like a reference leak; the `*new AST::ListValue` is not adopted.

Comment (alimpfard) [reply Shell/main.cpp:190]: Oh oops!

Comment (alimpfard) [reply Shell/main.cpp:190]: (there are quite a few around, I think, I'll search for them and fix them along with this one)

Comment (awesomekling) [reply Shell/main.cpp:190]: A `create()` helper that always returns a `NonnullRefPtr` would be nice here :)

File: Shell/AST.cpp [hunk 2]
```diff
@@ -1116,8 +1116,8 @@ RefPtr<Value> Pipe::run(RefPtr<Shell> shell)
 
     auto pipe_write_end = new FdRedirection(STDIN_FILENO, -1, Rewiring::Close::Destination);
     auto pipe_read_end = new FdRedirection(STDOUT_FILENO, -1, pipe_write_end, Rewiring::Close::RefreshDestination);
-    first_in_right.redirections.append(*pipe_write_end);
-    last_in_left.redirections.append(*pipe_read_end);
+    first_in_right.redirections.empend(adopt(static_cast<Redirection&>(*pipe_write_end)));
+    last_in_left.redirections.empend(adopt(static_cast<Redirection&>(*pipe_read_end)));
```
Comment (awesomekling) [inline Shell/AST.cpp:1120]: Please wrap the `new` in `adopt` directly instead of doing it later.

File: Shell/AST.h [hunk 3]
```diff
@@ -122,7 +122,7 @@ struct PathRedirection : public Redirection {
 };
 
 struct FdRedirection : public Redirection
-    , public Rewiring {
+    , virtual Rewiring {
```
Comment (awesomekling) [inline Shell/AST.h:125]: Wait, what's even going on here..
`Redirection` and `Rewiring` are both `RefCounted`. That's no good, that gives us two reference counts in the same object. That needs to be fixed.

Comment (alimpfard) [reply Shell/AST.h:125]: That's what I'm trying right now.
Not sure what I was doing when I came up with that idea.

File: Shell/AST.cpp (hunk 2 above)
Comment (alimpfard) [reply Shell/AST.cpp:1120]: With the double-`RefCounted` weirdness now fixed, this `static_cast` isn't needed anymore.

**Pull request merged**